docs/rfc/012-record-transaction-model.md

```diff
@@ -0,0 +1,243 @@
+Rather, the blockchain has to verify a zero-knowledge proof asserting that
+there exist values in `Input`, known to the creator of the transaction,
+such that the execution of the Leo program's `main`
+on those values and on the old records
+yields the new records;
+this is, roughly speaking, the assertion proved in zero-knowledge.
+
+### Proposed Leo Program Execution Model
+
+The current model described above seems adequate overall, but we need to:
+1. Clarify how Leo code reads old records and writes new records.
+2. Generalize from one entry point (i.e. the `main` function) to multiple entry points, in line with the smart contract paradigm.
+
+Generalizing from one `main` entry point to multiple ones is conceptually easy.
+It means that, instead of implicitly designating `main` as the only entry point,
+we need a mechanism to explicitly designate one or more Leo functions as entry points.
+
+A simple approach could be to use an annotation like `@entrypoint` to designate _entry point functions_:
+```
+@entrypoint
+function mint(...) -> ... { ... }
+
+@entrypoint
+function transfer(...) -> ... { ... }
+```
+This has a precedent, in the use of `@test` to designate Leo test functions that are not compiled to circuits.
+
+Another approach is to use a keyword, e.g.
+```
+entrypoint function mint(...) -> ... { ... }
+
+entrypoint function transfer(...) -> ... { ... }
+```
+Yet another approach is to group entrypoint functions inside a new block construct, e.g.
+```
+entrypoint {
+
+    function mint(...) -> ... { ... }
+
+    function transfer(...) -> ... { ... }
+}
+```
+
+Now let us turn to the issue of clarifying how the Leo code reads old records and writes new records.
+
+Given that records have a fixed structure with typed slots,
+their format could be described by a Leo circuit type,
+whose member variables correspond to the slots.
+The types of the slots would be fairly low-level,
+i.e. byte arrays (e.g. `u8[128]` for the payload)
+and unsigned integers (e.g. `u64` for the balance),
+because they must have a clear correspondence with the serialized form of records.
+This means that the Leo code may have to do
+its own deserialization of the payload bytes into higher-level Leo values;
+standard serialization/deserialization libraries for Leo types may be provided for this,
+as an independent and more generally useful feature.
+
+It may make sense to have a circuit type for the special `input` variable,
+which includes two slots for the two old records.
+All these circuit types should be explicitly documented,
+and available to the Leo program.
+
+However, we probably want `input` to be read-only,
+i.e. disallow assigning to an old record slot.
+Designating `input` as `const` does not seem right,
+as that designation normally means that it is compiled into the circuit.
+Instead, we could provide read-only access via member function (e.g. `get_payload()`, `get_balance()`),
```

Comment:
Having privacy (`pub` vs. private by default) sounds great. `get_payload` is also a pretty clean approach, but I'd suggest just having `payload()` instead, borrowing Rust naming conventions.

Reply:
Thanks, I just pushed a commit that renames the example functions as you suggest.